### PR TITLE
bug/table foreign ref

### LINKED ISSUE #713: reference error when writing units table
## 1) Bug

I am having trouble referencing the electrodes table in the units table:

### Steps to Reproduce
```python
from datetime import datetime

import numpy as np
import pandas as pd
from pynwb import NWBFile, NWBHDF5IO

nwbfile = NWBFile('session description', 'session identifier',
                  datetime.now().astimezone(), institution='UCSF',
                  lab='Chang Lab')

devices = ['a', 'b']
locations = ['a location', 'b location']
groups = []
for device_name, location in zip(devices, locations):
    # Create devices
    device = nwbfile.create_device(device_name)

    # Create electrode groups
    electrode_group = nwbfile.create_electrode_group(
        name=device_name + '_electrodes',
        description=device_name,
        location=location,
        device=device)
    groups.append(electrode_group)


electrodes_df = pd.DataFrame(
    {'location': ['c', 'c'],
     'group': groups,
     'x': [np.nan] * 2,
     'y': [np.nan] * 2,
     'z': [np.nan] * 2,
     'imp': [np.nan] * 2,
     'filtering': ['none'] * 2}
)

for _, row in electrodes_df.iterrows():
    nwbfile.add_electrode(**{label: row[label] for label in electrodes_df})

single_electrode_regions = [
    nwbfile.create_electrode_table_region([i], 'electrode i')
    for i in range(len(electrodes_df))]

all_spike_times = [[1., 2., 3., 4.],
                   [2., 3., 4.]]

for spike_times, single_electrode in zip(all_spike_times, single_electrode_regions):
    nwbfile.add_unit(spike_times=spike_times, electrode=single_electrode)


with NWBHDF5IO('test_file.nwb', 'w') as io:
    io.write(nwbfile)
```

```
Traceback (most recent call last):
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 426, in __add_refs
    call()
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 729, in _filler
    refs.append(self.__get_ref(item))
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/utils.py", line 375, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 902, in __get_ref
    return self.__file[path].ref
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/Users/bendichter/anaconda3/envs/dev_pynwb/lib/python3.6/site-packages/h5py/_hl/group.py", line 167, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 190, in h5py.h5o.open
KeyError: "Unable to open object (object 'electrodes' doesn't exist)"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/bendichter/dev/to_nwb/test_units.py", line 55, in <module>
    io.write(nwbfile)
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/utils.py", line 375, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 195, in write
    call_docval_func(super(HDF5IO, self).write, kwargs)
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/utils.py", line 268, in call_docval_func
    return func(*fargs, **fkwargs)
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/utils.py", line 375, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/io.py", line 40, in write
    self.write_builder(f_builder, **kwargs)
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/utils.py", line 375, in func_call
    return func(self, **parsed['args'])
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 411, in write_builder
    self.__add_refs()
  File "/Users/bendichter/dev/pynwb/src/pynwb/form/backends/hdf5/h5tools.py", line 429, in __add_refs
    raise RuntimeError('Unable to resolve reference')
RuntimeError: Unable to resolve reference
```

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?

### PR DESCRIPTION
**Pull request by ajtritt**
## Motivation

fix #713 

## Checklist

- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR description clearly describes problem and the solution?
- [x] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

### CODE REVIEW

**Review by oruebel — DISMISSED** [review]

File: docs/gallery/domain/ecephys.py [hunk 1]
```diff
@@ -127,6 +127,19 @@
...
+# Associate electrodes with units
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# The :ref:`PyNWB Basics tutorial <basics>` demonstrates how to add data about units and specifying custom metadata
+# about units. As mentioned :ref:`here <units_fields_ref>`, there are some optional fields for units, one of these
+# is *electrodes*. This field takes the *id* for the electrode that the unit corresponds to. For example, if two units
+# were inferred from the first electrode (*id* = 0), you would be specify that like so:
+
+nwbfile.add_unit(id=1, electrode=0)
+nwbfile.add_unit(id=2, electrode=0)
```
Comment (bendichter) [inline docs/gallery/domain/ecephys.py]: this syntax is nice. Would it be the same for `electrodes`? And would you mind adding some tests as well?

Comment (t-b) [reply docs/gallery/domain/ecephys.py]: I guess we throw when the electrode with the given id does not exist? Or maybe accepting a electrode object would be better?

Comment (ajtritt) [reply docs/gallery/domain/ecephys.py]: @bendichter There is no electrodes field in the units table. Right now, you can only specify a one-to-one relationship. I’ll fix the typo in the doc. 

@t-b as is, the check will happen at write.

**Review by bendichter — CHANGES_REQUESTED** [review]
hold on let me check some things before you commit

**Comment by bendichter** [comment]
Oh my mistake, the term was changed to `electrode_group`:

https://github.com/NeurodataWithoutBorders/pynwb/blob/9f4105c0cf3aa6f368a1f6a5c02b20c21b969cde/src/pynwb/data/nwb.misc.yaml#L134-L140

and here:

https://github.com/NeurodataWithoutBorders/pynwb/blob/9f4105c0cf3aa6f368a1f6a5c02b20c21b969cde/src/pynwb/misc.py#L199-L214

If I am not mistaken, this allows a many-to-many relationship. Sometimes, e.g. Utah arrays, each spike is associated with a single electrode (though an electrode can have multiple spikes). With other devices, e.g. tetrodes, multiple electrodes record a spike from a single unit, which would be many-to-many, so we need to continue to support electrode table references that can handle more than one electrode.

**Comment by tjd2002** [comment]
> With other devices, e.g. tetrodes, multiple electrodes record a spike from a single unit, which would be many-to-many, so we need to continue to support electrode table references that can handle more than one electrode.

Yup, definitely. Going further, for many clustering algorithms, each unit is actually recorded across a possibly unique 'neighborhood' of a multi-channel probe. Currently we represent that by creating an electrodeGroup for each unit.

**Comment by ajtritt** [comment]
@bendichter 

> If I am not mistaken, this allows a many-to-many relationship.

You are not wrong in your interpretation of that line of code... it does suggest that you can store a many-to-many relationship between units and electrodes. But I changed that line in this PR, because it was incorrect to begin with. 

If we need to support many-to-many (which seems completely reasonable), we can create a new type of column that combines the functionality of VectorData and DynamicTableRegion. However, I prefer that be addressed separately. At this point, this PR is to fix the ability to express one-to-many relationships between units and electrodes.

**Comment by bendichter** [comment]
@ajtritt OK but fyi most of the ephys labs adopting NWB 2.0 require many-to-many relationships between units and electrodes so this solution only helps a subset of ephys labs

**Review by bendichter — DISMISSED** [review]

**Review by oruebel — DISMISSED** [review]

**Review by oruebel — APPROVED** [review]

**Pull request merged**